bin/wasm-node/javascript/src/client.ts

```diff
@@ -599,58 +410,55 @@ export function start(options?: ClientOptions): Client {
-      const chainAddedPromise: Promise<Chain> = new Promise((resolve, reject) => {
-        chainAddedPromiseResolve = resolve;
-        chainAddedPromiseReject = reject;
-      });
-
-      pendingConfirmations.push({
-        ty: 'chainAdded',
-        reject: chainAddedPromiseReject as unknown as (error: AddChainError) => void,
-        resolve: chainAddedPromiseResolve as unknown as (c: Chain) => void,
-        jsonRpcCallback: options.jsonRpcCallback,
-      });
-
-      postMessage(worker, {
-        ty: 'addChain',
-        chainSpec: options.chainSpec,
-        databaseContent: typeof options.databaseContent === 'string' ? options.databaseContent : "",
-        potentialRelayChains: potentialRelayChainsIds,
-        jsonRpcRunning: !!options.jsonRpcCallback,
-      });
-
-      return chainAddedPromise;
+      const outcome = await instance.addChain(options.chainSpec, typeof options.databaseContent === 'string' ? options.databaseContent : "", potentialRelayChainsIds, options.jsonRpcCallback);
+
+      if (!outcome.success)
+        throw new AddChainError(outcome.error);
+
+      const chainId = outcome.chainId;
+      const wasDestroyed = { destroyed: false };
+
+      // `expected` was pushed by the `addChain` method.
+      // Resolve the promise that `addChain` returned to the user.
+      const newChain: Chain = {
+        sendJsonRpc: (request) => {
+          if (alreadyDestroyedError)
+            throw alreadyDestroyedError;
+          if (wasDestroyed.destroyed)
+            throw new AlreadyDestroyedError();
+          if (!options.jsonRpcCallback)
+            throw new JsonRpcDisabledError();
+          if (request.length >= 8 * 1024 * 1024)
+            return;
+          instance.request(request, chainId);
+        },
+        databaseContent: (maxUtf8BytesSize) => {
+          if (alreadyDestroyedError)
+            return Promise.reject(alreadyDestroyedError);
+          if (wasDestroyed.destroyed)
+            throw new AlreadyDestroyedError();
+          return instance.databaseContent(chainId, maxUtf8BytesSize);
+        },
+        remove: () => {
+          if (alreadyDestroyedError)
+            throw alreadyDestroyedError;
+          if (wasDestroyed.destroyed)
+            throw new AlreadyDestroyedError();
+          wasDestroyed.destroyed = true;
+          console.assert(chainIds.has(newChain));
+          chainIds.delete(newChain);
+          instance.removeChain(chainId);
+        },
+      };
+
+      chainIds.set(newChain, chainId);
+      return newChain;
     },
-    terminate: () => {
-      if (workerError)
-        return Promise.reject(workerError)
-      workerError = new AlreadyDestroyedError();
-
-      if (livenessTimeout !== null)
-        globalThis.clearTimeout(livenessTimeout)
-
-      return workerTerminate(worker)
+    terminate: async () => {
+      if (alreadyDestroyedError)
+        throw alreadyDestroyedError
+      alreadyDestroyedError = new AlreadyDestroyedError();
```

Comment:
Im not sure why this is needed

Reply:
It enforces that the API be used properly.
The API says "after having called terminate you can't call any other function anymore". This code actually enforces that.